icicle/include/hash/hash.cuh

```diff
@@ -0,0 +1,257 @@
+   * @param out pointer for squeeze results. Can be equal to states to do in-place squeeze
+   *
+   * @tparam S Type of the state element
+   */
+  template <typename S>
+  __global__ void generic_squeeze_states_kernel(
+    const S* states, unsigned int number_of_states, unsigned int width, unsigned int rate, unsigned int offset, S* out)
+  {
+    int idx = (blockIdx.x * blockDim.x) + threadIdx.x;
+    if (idx >= number_of_states) { return; }
+
+    for (int i = 0; i < rate; i++) {
+      if (states == out) {
+        S element = states[idx * width + offset];
+        __syncthreads();
+        out[idx * rate + i] = element;
+      } else {
+        out[idx * rate + i] = states[idx * width + offset];
+      }
+    }
+  }
+
+  template <typename Image>
+  class Hash
+  {
+  public:
+    unsigned int width;
+    unsigned int preimage_max_length;
+
+    Hash() = default;
+    Hash(unsigned int width, unsigned int preimage_max_length) : width(width), preimage_max_length(preimage_max_length)
+    {
+    }
+
+    virtual cudaError_t pad_many(
+      Image* states,
+      unsigned int number_of_states,
+      unsigned int input_block_len,
+      unsigned int rate,
+      const device_context::DeviceContext& ctx) const
+    {
+      return cudaError_t::cudaSuccess;
+    };
+
+    virtual cudaError_t squeeze_states(
+      const Image* states,
+      unsigned int number_of_states,
+      unsigned int rate,
+      unsigned int offset,
+      bool align,
+      Image* output,
+      const device_context::DeviceContext& ctx) const = 0;
+
+    virtual cudaError_t run_permutation_kernel(
+      const Image* states, Image* output, unsigned int number_of_states, bool aligned, const device_context::DeviceContext& ctx) const = 0;
+  };
+
+  struct SpongeConfig {
+    device_context::DeviceContext ctx;
+    bool are_inputs_on_device;
+    bool are_outputs_on_device;
+    unsigned int input_rate;
+    unsigned int output_rate;
+    unsigned int offset;
+    bool recursive_squeeze;
+    bool aligned;
+    bool is_async;
+  };
+
+  /**
+   * A function that returns the default value of [SpongeConfig](@ref SpongeConfig) for the [SpongeHasher](@ref
+   * SpongeHasher) class.
+   * @return Default value of [SpongeConfig](@ref SpongeConfig).
+   */
+  static SpongeConfig
+  default_sponge_config(const device_context::DeviceContext& ctx = device_context::get_default_device_context())
+  {
+    SpongeConfig config = {
+      ctx,   // ctx
+      false, // are_inputs_on_device
+      false, // are_outputs_on_device
+      0,     // input_rate
+      0,     // output_rate
+      0,     // offset
+      false, // recursive_squeeze
+      false, // recursive_squeeze
+      false, // is_async
+    };
+    return config;
+  }
+
+  template <typename H, typename PreImage, typename Image>
+  class SpongeHasher
+  {
+  public:
+    unsigned int get_width() const {
+      return static_cast<const H*>(this)->width;
```

Comment:
implies that H must derive from SpongeHasher?

Reply:
yes. This is the worst place from design standpoint:) I tried to replicate trait requirement from Rust. Initially I've done it in pur c++ style but it was worse